File: users/wpcarro/scratch/facebook/moderate/parser.py
class Parser(object):
    def __init__(self, tokens):
        self.tokens = tokens
        self.i = 0

    def prev(self):
        return self.tokens[self.i - 1]

    def curr(self):
        return self.tokens[self.i]

    def next(self):
        return self.tokens[self.i + 1]

    def consume(self):
        if not self.exhausted():
            self.i += 1
            return self.prev()

    def match(self, xs):
        if not self.exhausted() and self.curr() in xs:
            self.consume()
            return True
        return False

    def expect(self, xs):
        if not self.match(xs):
            raise Exception("Expected token \"{}\" but received \"{}\"".format(xs, self.curr()))
        return self.prev()

    def expect_predicate(self, predicate):
        if predicate(self.curr()):
            return self.consume()
        raise Exception("Expected token \"{}\" to pass predicate, but it did not".format(self.curr()))

    def exhausted(self):
        return self.i >= len(self.tokens)

File: users/wpcarro/scratch/compiler/parser.ml
(****************************************************************************** 
 * Defines a generic parser class.
 ******************************************************************************)

open Vec

exception ParseError of string

type token = string
type state = { i : int; tokens : token vec }

class parser (tokens : token vec) =
  object (self)
    val tokens = tokens
    val mutable i = 0

    method advance = i <- i + 1
    method prev : token option = Vec.get (i - 1) tokens
    method curr : token option = Vec.get i tokens
    method next : token option = Vec.get (i + 1) tokens

    method consume : token option =
      match self#curr with
      | None -> None
      | Some x as res ->
          self#advance;
          res

    method expect (x : token) =
      match self#curr with
      | Some y when x = y -> self#advance
      | _ -> raise (ParseError (Printf.sprintf "Expected %s" x))

    method matches (x : token) : bool =
      match self#curr with
      | None -> false
      | Some y ->
          if x = y then
            begin
              self#advance;
              true
            end
          else false

    method exhausted : bool = i >= Vec.length tokens
  end
